File: leetcode/solutions/0049.GroupAnagrams/groupAnagrams_test.ts
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
import groupAnagrams from "./groupAnagrams.ts";

const sorter = (arr1: string[], arr2: string[]) => {
  arr1.sort();
  arr2.sort();

  if (arr1.length !== arr2.length) {
    return arr1.length - arr2.length;
  }

  for (const [i, str1] of arr1.entries()) {
    const str2 = arr2[i];

    for (let i = 0; i < str1.length; i++) {
      if (str1[i] !== str2[i]) {
        return str1.charCodeAt(i) - str2.charCodeAt(i);
      }
    }
  }

  return 0;
};

Deno.test("0049. Group Anagrams", () => {
  const result = groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]);
  result.sort(sorter);

  const expected = [["ate", "eat", "tea"], ["nat", "tan"], ["bat"]];
  expected.sort(sorter);

  assertEquals(result, expected);
});

File: leetcode/solutions/0049.GroupAnagrams/groupAnagrams_test.go
package solution

import (
	"sort"
	"strings"
	"testing"

	"github.com/google/go-cmp/cmp"
)

type sortStrings [][]string

func (s sortStrings) Len() int {
	return len(s)
}

func (s sortStrings) Less(i, j int) bool {
	strs1, strs2 := s[i], s[j]

	if len(strs1) != len(strs2) {
		return len(strs1) < len(strs2)
	}

	if len(strs1) == 0 {
		return false
	}

	for k := 0; k < len(strs1[k]); k++ {
		compare := strings.Compare(strs1[k], strs2[k])
		if compare != 0 {
			return compare < 0
		}
	}

	return false
}

func (s sortStrings) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func Test_0049_GroupAnagrams(t *testing.T) {
	t.Parallel()

	tests := []struct {
		in  []string
		out [][]string
	}{
		{
			[]string{"eat", "tea", "tan", "ate", "nat", "bat"},
			[][]string{
				[]string{"ate", "eat", "tea"},
				[]string{"nat", "tan"},
				[]string{"bat"},
			},
		},
	}

	for _, tt := range tests {
		tt := tt

		t.Run("Test", func(t *testing.T) {
			t.Parallel()

			got := groupAnagrams(tt.in)

			for _, arr := range got {
				sort.Strings(arr)
			}
			for _, arr := range tt.out {
				sort.Strings(arr)
			}

			sort.Sort(sortStrings(got))
			sort.Sort(sortStrings(tt.out))

			if !cmp.Equal(got, tt.out) {
				t.Errorf("got %#v want %#v", got, tt.out)
			}
		})
	}
}
